File: 02-two_pointers/00125-Valid_Palindrome/00125-valid_palindrome.java
// 125: Valid Palindrome
// https://leetcode.com/problems/valid-palindrome/

class Solution {
    // SOLUTON
    public boolean isPalindrome(String s) {
        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            while (!Character.isLetterOrDigit(s.charAt(l)) && l<r) l++;
            while (!Character.isLetterOrDigit(s.charAt(r)) && l<r) r--;
            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))) return false;
            
            l++;
            r--;
        }

        return true;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s = "race a car";

        // OUTPUT
        var result = o.isPalindrome(s);
        System.out.println(result);        
    }
}

File: 02-two_pointers/00125-Valid_Palindrome/00125-valid_palindrome.py
# 125: Valid Palindrome
# https://leetcode.com/problems/valid-palindrome/

class Solution:
    # SOLUTION
    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1

        while l<r:
            while not s[l].isalnum() and l<r: l+=1
            while not s[r].isalnum() and l<r: r-=1
            if s[l].lower() != s[r].lower(): return False

            l+=1
            r-=1

        return True



if __name__ == "__main__":
    o = Solution()

    # INPUT
    s: str = "race a car"

    # OUTPUT
    result = o.isPalindrome(s)
    print(result)